src/rust/ide/src/controller/searcher.rs

```diff
@@ -0,0 +1,265 @@
+//! This module contains all structures related to Searcher Controller.
+
+use crate::prelude::*;
+
+use crate::notification;
+
+use data::text::TextLocation;
+use enso_protocol::language_server;
+use flo_stream::Subscriber;
+
+
+
+// =======================
+// === Suggestion List ===
+// =======================
+
+/// A single suggestion on the Searcher suggestion list.
+#[derive(Clone,CloneRef,Debug,Eq,PartialEq)]
+pub enum Suggestion {
+    /// Suggestion for input completion: possible functions, arguments, etc.
+    Completion(Rc<model::suggestion_database::Entry>)
+    // In future, other suggestion types will be added (like suggestions of actions, etc.).
+}
+
+/// List of suggestions available in Searcher.
+#[derive(Clone,CloneRef,Debug)]
+pub enum Suggestions {
+    /// The suggestion list is still loading from the Language Server.
+    Loading,
+    /// The suggestion list is loaded.
+    #[allow(missing_docs)]
+    Loaded {
+        list : Rc<Vec<Suggestion>>
+    },
+    /// Loading suggestion list results in error.
+    Error(Rc<failure::Error>)
+}
+
+impl Suggestions {
+    /// Check if suggestion list is still loading.
+    pub fn is_loading(&self) -> bool {
+        match self {
+            Self::Loading => true,
+            _             => false,
+        }
```

Comment:
Why not just `self == Self::Loading` ?

Reply:
I cannot implement Eq for this structure, because it contains `failure::Error` which does not implement Eq.

Reply:
Oh, I did not notice it. Btw, thy we are using `fialure::Error` in the first place? Why we do not keep all possible errors just as enums?

Reply:
We do keep all errors as enums, but failure helps a lot if you have two error's kind and want to return both of them. And those error kinds are often just an implementation detail, which failure::Error elegantly hides.

However, this crate is deprecated and we have issue to replace it: https://github.com/enso-org/ide/issues/443